### PR TITLE
TAP:ESC motors and LED working

### PR DESCRIPTION
**Pull request by davids5**
@LorenzMeier, @bkueng 

The code is still hardcoded for 4 ESC, but the basics are working. 

N.B. To Stop the motors a value of 0 need to be sent @bkueng motor_test could benefit  from mask foe channels, a stop command, etc

 LED color mask are misleading - LED is On or Off per ESC.

### CODE REVIEW

File: src/drivers/tap_esc/drv_tap_esc.h [hunk 1]
```diff
@@ -59,49 +59,121 @@
...
 	uint8_t msg_id;
-	uint8_t data[100];
+	union {
+		InfoRequest 			reqInfo;
+		ConfigInfoBasicRequest 	reqConfigInfoBasic;
+		RunReq					reqRun;
+
+		ConfigInfoBasicResponse rspConfigInfoBasic;
+		RunInfoRepsonse			rspRunInfo;
+		uint8_t bytes[100];
```
Comment (davids5) [inline src/drivers/tap_esc/drv_tap_esc.h]: Could use to remove the length

File: src/drivers/tap_esc/drv_tap_esc.h [hunk 2]
```diff
@@ -140,6 +212,8 @@ typedef enum {
 	ESC_STATUS_ERROR_HARDWARE,
 	ESC_STATUS_ERROR_LOSE_PROPELLER,
 	ESC_STATUS_ERROR_OVER_CURRENT,
+	ESC_STATUS_ERROR_MOTOR_HIGH_SPEED_LOSE_STEP,
```
Comment (davids5) [inline src/drivers/tap_esc/drv_tap_esc.h]: Need updating
S/B

typedef enum {
ESC_STATUS_HEALTHY,
ESC_STATUS_WARNING_LOW_VOLTAGE,
ESC_STATUS_WARNING_OVER_HEAT,
ESC_STATUS_ERROR_MOTOR_LOW_SPEED_LOSE_STEP,
ESC_STATUS_ERROR_MOTOR_STALL,
ESC_STATUS_ERROR_HARDWARE,
ESC_STATUS_ERROR_LOSE_PROPELLER,
ESC_STATUS_ERROR_OVER_CURRENT,
ESC_STATUS_ERROR_LOSE_COMMAND,
} ESCBUS_ENUM_ESC_STATUS;

File: src/drivers/tap_esc/tap_esc.cpp [hunk 3]
```diff
@@ -603,13 +712,13 @@ TAP_ESC::cycle()
...
-				feed_back_data = *(ESC_FEEDBACK_DATA *)feed_back_packet.data;
-				_esc_feedback.esc[feed_back_data.ESC_ID].esc_rpm = feed_back_data.speed;
-				_esc_feedback.esc[feed_back_data.ESC_ID].esc_voltage = feed_back_data.voltage;
-				_esc_feedback.esc[feed_back_data.ESC_ID].esc_state = feed_back_data.ESC_STATUS;
-				_esc_feedback.esc[feed_back_data.ESC_ID].esc_vendor = esc_status_s::ESC_VENDOR_TAP;
+		if (parse_tap_esc_feedback(&uartbuf, &_packet) == true) {
+			if (_packet.msg_id == ESCBUS_MSG_ID_RUN_INFO) {
+				RunInfoRepsonse &feed_back_data = _packet.d.rspRunInfo;
+				_esc_feedback.esc[feed_back_data.channelID].esc_rpm = feed_back_data.speed;
+//				_esc_feedback.esc[feed_back_data.channelID].esc_voltage = feed_back_data.voltage;
```
Comment (davids5) [inline src/drivers/tap_esc/tap_esc.cpp]: Note Lost Voltage but have Current

File: src/drivers/tap_esc/tap_esc.cpp [hunk 4]
```diff
@@ -802,7 +912,7 @@ int tap_esc_start(void)
 
 	if (tap_esc == nullptr) {
 
-		tap_esc = new TAP_ESC;
+		tap_esc = new TAP_ESC(_supported_channel_count);
```
Comment (davids5) [inline src/drivers/tap_esc/tap_esc.cpp]: Refactor and pass param using argv and then to constructor. I am surprised the uart woks opened on caller thread and passed to new task.

File: src/drivers/tap_esc/tap_esc.cpp [hunk 5]
```diff
@@ -947,7 +1057,7 @@ void start()
 	_task_handle = px4_task_spawn_cmd("tap_esc_main",
 					  SCHED_DEFAULT,
 					  SCHED_PRIORITY_MAX,
-					  1500,
+					  2500,
```
Comment (davids5) [inline src/drivers/tap_esc/tap_esc.cpp]: May be to bug - need to examine

File: src/drivers/tap_esc/tap_esc.cpp (hunk 4 above)
Comment (bkueng) [inline src/drivers/tap_esc/tap_esc.cpp]: `initialize_uart` is called from `task_main`, which runs on the same thread as the driver main loop. So this is correct (and why it works), but the naming & startup sequence is really confusing.

**Comment by LorenzMeier** [comment]
@bkueng Are you able to spin propellers with this?

**Comment by bkueng** [comment]
yes, all of them work. however when I let them run faster (50% max), they start to make a cracking noise and don't run smoothly anymore. Needs to be investigated further.

**Comment by LorenzMeier** [comment]
Its possible they loose sync without the props. We should test with props attached.

**Comment by LorenzMeier** [comment]
Can we fix the CI test failure?

**Comment by bkueng** [comment]
The CI failure is something arch specific and @davids5 knows best how to resolve this.

**Comment by davids5** [comment]
Why is the tap_esc driver is in px4fmu-v4_default? The issues is that HW does not have the mux, hence the undefined GPIO_S0. If the driver is not need in the v4 config - removing it will solve the issue. If it is needed then Adding #if defined(GPIO_S0) wrapper on the muxing code will fix it.

**Comment by bkueng** [comment]
Because tap_esc can be used with eg. Pixracer as well, so it needs to stay in there. The ifdef is fine then, but maybe add a comment why it is needed.

**Comment by LorenzMeier** [comment]
Rebased and applied.

**Pull request closed without merging**